CBMC-6.0.0 fails if local variable name overloads name of quantified variable in contract

CBMC version: 6.0.0-preview (built 17th June 2024)
Operating system: macOS
Exact command line resulting in the issue: make TARGET=init_st
What behaviour did you expect: success
What happened instead:

CBMC fails on the latest examples here:
 https://github.com/rod-chapman/cbmc-examples/tree/main/arrays

In particular, see the init_st() function.

Attempt to verify with
 make TARGET=init_st

results in
```
rodchap@f4d4889dcf6d arrays % make TARGET=init_st
rm -f *.goto
goto-cc --function init_st_harness -DCBMC -o ar.goto ar.c
ar.c: In function 'init_st':
ar.c:171:1: error: symbol 'init_st::1::i' redefined with a different type:
Original: unsigned int
     New: size_t
     size_t i;
CONVERSION ERROR
make: *** [ar.goto] Error 1

```

This appears to be because the quantified variable "i" in the "ensures" contract (in ar.h) has the same name as a local variable "i" in the function body. This should not happen - I assume that quentified variables are supposed to have their own scope and possibly their own name-space.

Changing the name of the quantified variable so it doesn't clash appears to make this problem go away.

--- a/regression/ansi-c/contracts_scope1/main.c
+++ b/regression/ansi-c/contracts_scope1/main.c
@@ -1,0 +1,28 @@
+#define C 8
+
+typedef unsigned st[C];
+
+// clang-format off
+void init_st(st dst)
+  __CPROVER_requires(__CPROVER_is_fresh(dst, sizeof(st)))
+  __CPROVER_assigns(__CPROVER_object_whole(dst))
+  __CPROVER_ensures(__CPROVER_forall {
+    unsigned i; (0 <= i && i < C) ==> dst[i] == 0
+  });
+// clang-format on
+
+void init_st(st dst)
+{
+  __CPROVER_size_t i;
+  for(i = 0; i < C; i++)
+  {
+    dst[i] = 0;
+  }
+}
+
+int main()
+{
+  st dest;
+
+  init_st(dest);
+}

--- a/regression/ansi-c/contracts_scope1/test.desc
+++ b/regression/ansi-c/contracts_scope1/test.desc
@@ -1,0 +1,8 @@
+CORE
+main.c
+
+^EXIT=0$
+^SIGNAL=0$
+--
+^warning: ignoring
+^CONVERSION ERROR$

--- a/src/ansi-c/parser.y
+++ b/src/ansi-c/parser.y
@@ -3337,17 +3337,27 @@ parameter_abstract_declarator:
         ;
 
 cprover_function_contract:
-          TOK_CPROVER_ENSURES '(' ACSL_binding_expression ')'
+          TOK_CPROVER_ENSURES
+        {
+          PARSER.new_scope("ensures::");
+        }
+          '(' ACSL_binding_expression ')'
         {
           $$=$1;
           set($$, ID_C_spec_ensures);
-          mto($$, $3);
+          mto($$, $4);
+          PARSER.pop_scope();
         }
-        | TOK_CPROVER_REQUIRES '(' ACSL_binding_expression ')'
+        | TOK_CPROVER_REQUIRES
+        {
+          PARSER.new_scope("requires::");
+        }
+          '(' ACSL_binding_expression ')'
         {
           $$=$1;
           set($$, ID_C_spec_requires);
-          mto($$, $3);
+          mto($$, $4);
+          PARSER.pop_scope();
         }
         | cprover_contract_assigns
         | cprover_contract_frees

